SourceImage.h shouldn't include "ImageBuffer.h"

fujii opened the pull request:
#### 3e9a6fa4d46f2b5d79c4b56bfebf68155add517b
<pre>
SourceImage.h shouldn&apos;t include &quot;ImageBuffer.h&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=283541">https://bugs.webkit.org/show_bug.cgi?id=283541</a>

Reviewed by Ross Kirsling.

<a href="https://commits.webkit.org/286935@main">https://commits.webkit.org/286935@main</a> was a wrong fix. Reverted it.
And, define default constructors and operator= in the cpp file.

* Source/WebCore/platform/graphics/SourceImage.cpp:
* Source/WebCore/platform/graphics/SourceImage.h:

Canonical link: <a href="https://commits.webkit.org/286945@main">https://commits.webkit.org/286945@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae1a7d44a65f81ee16a59a5f7cf09da9fa53d7d7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77660 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/56693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30582 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/28938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79777 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65849 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4993 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/82285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/28938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80726 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/50814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/66643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/48204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24139 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/27269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/69318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/24483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83663 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/5040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/69081 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/5197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66611 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17070 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/4987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/5002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/8440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/6765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/WebCore/platform/graphics/SourceImage.cpp:
```diff
@@ -37,6 +37,10 @@ SourceImage::SourceImage(ImageVariant&& imageVariant)
 {
 }
 
+SourceImage::SourceImage(const SourceImage&) = default;
+SourceImage::SourceImage(SourceImage&&) = default;
+SourceImage& SourceImage::operator=(const SourceImage&) = default;
+SourceImage& SourceImage::operator=(SourceImage&&) = default;
```
Seems like these could just be kept in the header, but I guess the destructor's already here.

fujii replied:
I don't know any WebKit documents. But, this is a common C++ practice which is explained in C++ books.

rkirsling replied:
I know setting `default` for constructors and copy constructors is a thing, but I didn't expect it to be a solution for the problem of `SourceImage` having a `std::variant<..., Ref<ImageBuffer>, ...>` member variable without an `ImageBuffer.h` include. What's the connection?

fujii replied:
generated SourceImage's member functios call ImageBuffer's member functions.
* ImageBuffer::ImageBuffer
* ImageBuffer::~ImageBuffer
* ImageBuffer::operator=

That's the reason ImageBuffer.h has to be incluced.

rkirsling reviewed (approved):
Huh, wild. Should this be spelled out in the documentation for `Ref` somehow?

fujii committed:
SourceImage.h shouldn't include "ImageBuffer.h"
https://bugs.webkit.org/show_bug.cgi?id=283541

Reviewed by Ross Kirsling.

https://commits.webkit.org/286935@main was a wrong fix. Reverted it.
And, define default constructors and operator= in the cpp file.

* Source/WebCore/platform/graphics/SourceImage.cpp:
* Source/WebCore/platform/graphics/SourceImage.h:

Canonical link: https://commits.webkit.org/286945@main
